0012-common-ledger-api/0012-common-ledger-api.md

```diff
@@ -0,0 +1,1201 @@
+
+**Authorization:** The `from` field of the message MUST match the account owner. Unauthenticated users MUST NOT be able to send messages. A ledger MAY allow admin connections to send messages whose `from` value matches the ledger's base [URL][].
+
+#### Request Format
+
+POST a JSON [Message resource][] to the `message` [URL][].
+
+**Metadata URL Name:** `message`. Suggested generic path: `message`
+
+#### Response Format
+
+A successful response uses the HTTP response code **204 No Content** and contains no message body.
+
+#### Example
+
+Request:
+
+```json
+POST /messages
+Content-Type: application/json
+
+{
+  "from": "https://blue.ilpdemo.ripple.com/ledger/accounts/bob",
+  "to": "https://blue.ilpdemo.ripple.com/ledger/accounts/alice",
+  "ledger": "https://blue.ilpdemo.ripple.com/ledger",
+  "data": {
+    "method": "quote_request",
+    "id": "721e4126-98a1-4974-b35a-8a8f4655f934",
+    "data": {
+      "source_amount": "100.25",
+      "source_address": "example.eur-ledger.alice",
+      "destination_address": "example.usd-ledger.bob",
+      "source_expiry_duration": "6000",
+      "destination_expiry_duration": "5"
+    }
+  }
+}
+```
+
+**Note:** This example uses the [Interledger Quoting Protocol](https://github.com/interledger/rfcs/blob/master/0008-interledger-quoting-protocol/0008-interledger-quoting-protocol.md) quote request type as example data. However, the `data` field may contain arbitrary objects, so the ledger MUST NOT require a specific format for such data.
+
+
+Response:
+
+```
+HTTP/1.1 204 No Content
+```
+
+#### Errors
+
+- [InvalidBodyError][]
+- [InvalidUriParameterError][]
+- [UnprocessableEntityError][]
+
+
+### Get Auth Token
+[Get Auth Token]: #get-auth-token
+
+Get a token that can be used to authenticate future requests.
+
+**Note:** This method is REQUIRED for ledgers to authenticate [WebSocket][] connections. If the ledger does not authenticate WebSocket connections, this method is OPTIONAL. The ledger MAY allow clients to authenticate for any other methods of the API using the tokens returned by this method.
+
+#### Request Format
+
+GET the `auth_token` [URL][].
+
+Depending on the [authentication mechanism](#authorization-and-authentication) used by this ledger, the ledger MAY require the HTTP `Auth` header with a valid username and password, or the ledger may require a different method of authentication.
+
+**Metadata URL Name:** `auth_token`.
+
+#### Response Format
+
+A successful response uses the HTTP response code **200 OK** and contains a JSON object with a token that can be used for subsequent requests. The ledger can generate the token with [JWT](https://jwt.io/) or any similar system.
+
+#### Example
+
+Request:
+
+```
+GET /auth_token
+Authorization: Basic myUsername:securePassphrase
+```
+
+Response:
+
+```json
+HTTP/1.1 200 OK
+Content-Type: application/json
+
+{
+  "token": "9AtVZPN3t49Kx07stO813UHXv6pcES"
+}
+```
+
+#### Errors
+
+- [Unauthorized][]
+
+
+
+
+## WebSocket
```

Comment:
@sappenin and @earizon I think the core Java stuff is stabalizing to the point that we should definitely try a second adaptor implementation.

Could we try implementing a gRPC based interface between the client and your ledger?

You'd implement the org.interledger.ilp.ledger.LedgerAdaptor interface and then should be able to simply instantiate a Client with your adaptor instead of the REST adaptor.